libtailscale/backend.go

```diff
@@ -346,3 +332,29 @@ func (a *App) newBackend(dataDir, directFileRoot string, appCtx AppContext, stor
 	}()
 	return b, nil
 }
+
+func (b *backend) isConfigNonNilAndDifferent(rcfg *router.Config, dcfg *dns.OSConfig) bool {
+	if reflect.DeepEqual(rcfg, b.lastCfg) && reflect.DeepEqual(dcfg, b.lastDNSCfg) {
+		b.logger.Logf("isConfigNonNilAndDifferent: no change to Routes or DNS, ignore")
+		return false
+	}
+	return rcfg != nil
+}
+
+func (a *App) closeVpnService(err error, b *backend) {
+	log.Printf("VPN update failed: %v", err)
+
+	mp := new(ipn.MaskedPrefs)
+	mp.WantRunning = false
+	mp.WantRunningSet = true
+
+	_, localApiErr := a.EditPrefs(*mp)
+	if err != nil {
+		log.Printf("localapi edit prefs error %v", localApiErr)
+	}
```

Comment:
I think it should be either 
```suggestion
	_, err = a.EditPrefs(*mp)
	if err != nil {
		log.Printf("localapi edit prefs error %v", err)
	}
```
or
```suggestion
	_, localApiErr := a.EditPrefs(*mp)
	if localApiErr != nil {
		log.Printf("localapi edit prefs error %v", localApiErr)
	}
```

Personally, I'd slightly prefer the first option or perhaps a variant of it with a short init statement in the `if`.

Reply:
This doesn't seem to be fixed yet, and I apologize for any confusion.
My main point, though, was that we should check whether the error returned from `EditPrefs` is non-nil.

Both the current and original code store the error in localApiErr, but then check `err != nil` instead of `localApiErr != nil`.

Whether to re-use the existing `err` variable or introduce `localApiErr`, and whether to use a short init statement, is secondary.

Reply:
my mistake! I was not typing what I was thinking!